cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsIDBrokerObjectStorageValidator.java

```diff
@@ -0,0 +1,258 @@
+package com.sequenceiq.cloudbreak.cloud.aws;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.amazonaws.auth.policy.Action;
+import com.amazonaws.auth.policy.Policy;
+import com.amazonaws.auth.policy.Resource;
+import com.amazonaws.auth.policy.Statement;
+import com.amazonaws.services.identitymanagement.AmazonIdentityManagement;
+import com.amazonaws.services.identitymanagement.model.EvaluationResult;
+import com.amazonaws.services.identitymanagement.model.GetInstanceProfileRequest;
+import com.amazonaws.services.identitymanagement.model.InstanceProfile;
+import com.amazonaws.services.identitymanagement.model.NoSuchEntityException;
+import com.amazonaws.services.identitymanagement.model.Role;
+import com.amazonaws.services.identitymanagement.model.ServiceFailureException;
+import com.amazonaws.services.identitymanagement.model.SimulatePrincipalPolicyRequest;
+import com.amazonaws.services.identitymanagement.model.SimulatePrincipalPolicyResult;
+import com.sequenceiq.cloudbreak.cloud.exception.CloudConnectorException;
+import com.sequenceiq.cloudbreak.cloud.model.filesystem.CloudFileSystemView;
+import com.sequenceiq.cloudbreak.cloud.model.filesystem.CloudS3View;
+import com.sequenceiq.common.api.cloudstorage.AccountMappingBase;
+import com.sequenceiq.common.api.cloudstorage.StorageLocationBase;
+import com.sequenceiq.common.model.CloudIdentityType;
+import com.sequenceiq.common.model.CloudStorageCdpService;
+
+public class AwsIDBrokerObjectStorageValidator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsIDBrokerObjectStorageValidator.class);
+
+    private final AmazonIdentityManagement iam;
+
+    public AwsIDBrokerObjectStorageValidator(AmazonIdentityManagement amazonIdentityManagement) {
+        this.iam = amazonIdentityManagement;
+    }
+
+    public void validateObjectStorage(List<CloudFileSystemView> cloudFileSystems) {
+        for (CloudFileSystemView cloudFileSystemView : cloudFileSystems) {
+            CloudS3View cloudFileSystem = (CloudS3View) cloudFileSystemView;
+            InstanceProfile instanceProfile = getInstanceProfile(cloudFileSystem.getInstanceProfile());
+            if (instanceProfile != null) {
+                CloudIdentityType cloudIdentityType = cloudFileSystem.getCloudIdentityType();
+                if (CloudIdentityType.ID_BROKER.equals(cloudIdentityType)) {
+                    checkInstanceProfileTrust(instanceProfile);
+                    checkIDBrokerAssumeRolePermissions(instanceProfile, cloudFileSystem);
```

Comment:
A lot of permission checks are done here and if any of them fails an exception is thrown. However there might be other cases of insufficient permissions, and the user won't be notified about those, because an exception was thrown at the first failure. To mitigate this I suggest to use the com.sequenceiq.cloudbreak.validation.ValidationResult and it's builder inner class to collect all the errors, and at the end of this method this contains anything, then an exception can be thrown - now containing all the errors that this validator can find.

Reply:
Yea I had seen this used elsewhere. I will use that to collect all the errors up front.

Reply:
Latest push collects validation results all at once.